File: solutions/typescript/0095.ts
/**
 * 95. 不同的二叉搜索树 II
 * 区间递归：generate(start, end) 枚举 [start, end] 内每个值作为根，
 * 左子树取 generate(start, root-1)，右子树取 generate(root+1, end)，
 * 两两组合得到该区间内所有可能的 BST。
 *
 * 注：TreeNode 类型由 LeetCode 平台注入，本文件用 untyped TNode 接口避免冲突。
 */

interface TNode { val: number; left: TNode | null; right: TNode | null }

function generateTrees(n: number): (TNode | null)[] {
    if (n === 0) return [];

    const generate = (start: number, end: number): (TNode | null)[] => {
        if (start > end) return [null];
        const result: (TNode | null)[] = [];
        for (let rootVal = start; rootVal <= end; rootVal++) {
            const lefts = generate(start, rootVal - 1);
            const rights = generate(rootVal + 1, end);
            for (const left of lefts) {
                for (const right of rights) {
                    result.push({ val: rootVal, left, right });
                }
            }
        }
        return result;
    };

    return generate(1, n);
}

export default generateTrees;

File: solutions/rust/0095.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
use std::cell::RefCell;
use std::rc::Rc;

type Link = Option<Rc<RefCell<TreeNode>>>;

/// 95. 不同的二叉搜索树 II
/// 区间递归：generate(start, end) 枚举 [start, end] 内每个值作为根，
/// 左子树取 generate(start, root-1)，右子树取 generate(root+1, end)，
/// 两两组合得到该区间内所有可能的 BST。
pub fn generate_trees(n: i32) -> Vec<Link> {
    if n == 0 {
        return Vec::new();
    }

    fn generate(start: i32, end: i32) -> Vec<Link> {
        if start > end {
            return vec![None];
        }
        let mut result = Vec::new();
        for root_val in start..=end {
            let lefts = generate(start, root_val - 1);
            let rights = generate(root_val + 1, end);
            for left in &lefts {
                for right in &rights {
                    result.push(Some(Rc::new(RefCell::new(TreeNode {
                        val: root_val,
                        left: left.clone(),
                        right: right.clone(),
                    }))));
                }
            }
        }
        result
    }

    generate(1, n)
}

impl Solution {
    pub fn generate_trees(n: i32) -> Vec<Link> {
        generate_trees(n)
    }
}
